tests/cancun/eip5656_mcopy/test_mcopy.py

```diff
@@ -0,0 +1,218 @@
+
+    # If the memory was extended beyond the initial range, store the last word of the resulting
+    # memory into storage too
+    if len(final_memory) > len(initial_memory):
+        last_word = ceiling_division(len(final_memory), 0x20) - 1
+        bytecode += Op.SSTORE(last_word, Op.MLOAD(last_word * 0x20))
+        storage[last_word] = final_memory[last_word * 0x20 : (last_word + 1) * 0x20].ljust(
+            32, b"\x00"
+        )
+
+    # We could also keccak the memory and store the hash in storage for good measure ?
+
+    return (bytecode, storage)
+
+
+@pytest.fixture
+def pre(  # noqa: D103
+    bytecode_storage: Tuple[bytes, Storage.StorageDictType]
+) -> Mapping[str, Account]:
+    return {
+        TestAddress: Account(balance=10**40),
+        code_address: Account(code=bytecode_storage[0]),
+    }
+
+
+@pytest.fixture
+def tx(dest: int, src: int, length: int) -> Transaction:  # noqa: D103
+    return Transaction(
+        to=code_address,
+        data=to_hash_bytes(dest) + to_hash_bytes(src) + to_hash_bytes(length),
+        gas_limit=1_000_000,
+    )
+
+
+@pytest.fixture
+def post(  # noqa: D103
+    bytecode_storage: Tuple[bytes, Storage.StorageDictType]
+) -> Mapping[str, Account]:
+    return {
+        code_address: Account(storage=bytecode_storage[1]),
+    }
+
+
+@pytest.mark.parametrize(
+    "dest,src,length",
+    [
+        (0x00, 0x00, 0x00),
+        (2**256 - 1, 0x00, 0x00),
+        (0x00, 0x00, 0x01),
+        (0x00, 0x00, 0x20),
+        (0x01, 0x00, 0x01),
+        (0x01, 0x00, 0x20),
+        (0x11, 0x11, 0x01),
+        (0x11, 0x11, 0x20),
+        (0x11, 0x11, 0x40),
+        (0x10, 0x00, 0x40),
+        (0x00, 0x10, 0x40),
+        (0x0F, 0x10, 0x40),
+        (0x100, 0x01, 0x01),
+        (0x100, 0x01, 0x20),
+        (0x100, 0x01, 0x1F),
+        (0x100, 0x01, 0x21),
+        (0x00, 0x00, 0x100),
+        (0x100, 0x00, 0x100),
+        (0x200, 0x00, 0x100),
+        (0x00, 0x100, 0x100),
+        (0x100, 0x100, 0x01),
+    ],
+    ids=[
+        "zero_inputs",
+        "zero_length_out_of_bounds_destination",
+        "single_byte_rewrite",
+        "full_word_rewrite",
+        "single_byte_forward_overwrite",
+        "full_word_forward_overwrite",
+        "mid_word_single_byte_rewrite",
+        "mid_word_single_word_rewrite",
+        "mid_word_multi_word_rewrite",
+        "two_words_forward_overwrite",
+        "two_words_backward_overwrite",
+        "two_words_backward_overwrite_single_byte_offset",
+        "single_byte_memory_extension",
+        "single_word_memory_extension",
+        "single_word_minus_one_byte_memory_extension",
+        "single_word_plus_one_byte_memory_extension",
+        "full_memory_rewrite",
+        "full_memory_copy",
+        "full_memory_copy_offset",
+        "full_memory_clean",
+        "out_of_bounds_memory_extension",
+    ],
+)
+@pytest.mark.valid_from("Cancun")
+def test_valid_mcopy_operations(
```

Comment:
We could maybe add a separate test case similar to this for invalid parameters - `test_mcopy_invalid_params` only if we are able to generate them or if you think they are useful:
```python
(2**256, 2**256, 0x00),  # src_and_dest_out_of_bounds
(2**256, 0x00, 0x00),  # src_out_of_bounds
...
(0x00, 0x00, -0x01),  # negative_length
```

Reply:
I don't think there's a way to cover the first two since the EVM stack only reaches `2**256 - 1`, so even using these as inputs for the transaction data, by the time it gets converted into the EVM, we will lose the range.

The last one is covered by `test_mcopy_huge_memory_expansion` (`-1` is equivalent to `2**256 - 1`)